### DISCUSSION TITLE
如何改引用文献的人名样式？

### DISCUSSION THREAD

**Discussion by iTomxy** [sikouhjw/gdutthesis#66] [Q&A] [answered]
模板中参考文献的作者名是全部大写（形如 `ZHANG S`），大老板要求改成「首字母大写、其它小写」的样式（形如 `Zhang San`），请问要如何设置？

我在 texlive 的安装路径中找到一个跟 gb7714-2015 相关的说明文件：*E:/texlive/2021/texmf-dist/doc/latex/biblatex-gb7714-2015/biblatex-gb7714-2015.pdf*，里面讲到可以设置 `biblatex` 宏包的 `gbnamefmt` 参数，将其设为 `gbnamefmt = quanpin` 时就是对应上述那种样式。于是我尝试：

- 在 `\__gdut_biblatex_pre_setup`<sup>[1]</sup> 中，把 `gbnamefmt = quanpin` 加进末尾句 `\__gdut_pass_options_to_biblatex:n` 中；
- 在 `\__gdut_biblatex_pre_setup`<sup>[1]</sup> 末尾另加一句 `\PassOptionsToPackage { gbnamefmt = quanpin } { biblatex }`；
- 在 `\__gdut_bibtex_setup`<sup>[2]</sup> 中，在那句 `\cs_set:Npn \bibsection { \chapter{参考文献}{References} }` 后面加一句 `\__gdut_pass_options_to_biblatex:n { gbnamefmt = quanpin }`；

三种方法都**没**效果，将 `gbnamefmt` 设为其它参数也没有效果。

[3] 有讲一种修改 .bst 文件达到这个效果的方法，gdutthesis.pdf<sup>[5]</sup> 中提到 `stype/bib-style` 选 `〈其他样式〉` 且 `bib-backend = bibtex` 时要保证*相应的 .bst 文件* 能被调用。所以正解是把参考文献的格式改为用 .bst 文件控制吗？如果这样的话要怎么操作呢？（比如 `stype/bib-style` 要设为什么）

---

1. [\\__gdut_biblatex_pre_setup](https://github.com/sikouhjw/gdutthesis/blob/main/gdutthesis.cls#L1465)
2. [\\__gdut_bibtex_setup](https://github.com/sikouhjw/gdutthesis/blob/main/gdutthesis.cls#L1435)
3. [Latex 参考文献格式GB/7714——2005如何将英文作者名自动大写改成普通的写法](https://blog.csdn.net/X_And_Y/article/details/104867559)
4. [在latex中使用GB/T 7714—2015参考文献著录规则](https://blog.csdn.net/G_Barble/article/details/111499448)
5. [gdutthesis.pdf](https://github.com/sikouhjw/gdutthesis/releases/latest/download/gdutthesis.pdf)

**Comment by sikouhjw** [comment, marked as answer]
> 模板中参考文献的作者名是全部大写（形如 `ZHANG S`），大老板要求改成「首字母大写、其它小写」的样式（形如 `Zhang San`），请问要如何设置？

首先全大写是国标的要求，当然，`texdoc gbt7714` 和 `texdoc biblatex-gb7714-2015` 都提供了相应的接口。

> 我在 texlive 的安装路径中找到一个跟 gb7714-2015 相关的说明文件：_E:/texlive/2021/texmf-dist/doc/latex/biblatex-gb7714-2015/biblatex-gb7714-2015.pdf_，里面讲到可以设置 `biblatex` 宏包的 `gbnamefmt` 参数，将其设为 `gbnamefmt = quanpin` 时就是对应上述那种样式。于是我尝试：
> 
>     * 在 `\__gdut_biblatex_pre_setup`[1] 中，把 `gbnamefmt = quanpin` 加进末尾句 `\__gdut_pass_options_to_biblatex:n` 中；
> 
>     * 在 `\__gdut_biblatex_pre_setup`[1] 末尾另加一句 `\PassOptionsToPackage { gbnamefmt = quanpin } { biblatex }`；
> 
>     * 在 `\__gdut_bibtex_setup`[2] 中，在那句 `\cs_set:Npn \bibsection { \chapter{参考文献}{References} }` 后面加一句 `\__gdut_pass_options_to_biblatex:n { gbnamefmt = quanpin }`；
> 
> 
> 三种方法都**没**效果，将 `gbnamefmt` 设为其它参数也没有效果。

你找的地方是对的，但可能你需要的是 `gbnamefmt=lowercase`（大小写由输入信息确定不做改变），你写的代码也是对的，没效果可能是因为『辅助文件没有删除』，导致 latexmk 认为你没有修改参考文献，所以没有重新编译；也可能是因为你没有选择 `bib-backend = {biblatex}`，导致没有使用 biblatex。

按以下步骤修改：
1. 删除全部辅助文件
2. bib 后端设置为 `bib-backend = {biblatex}`
3. 在导言区添加 `\PassOptionsToPackage{gbnamefmt=lowercase}{biblatex}`
4. 用 `latexmk -xelatex` 或者 `xelatex biber xelatex xelatex` 编译

> [3] 有讲一种修改 .bst 文件达到这个效果的方法，gdutthesis.pdf[5] 中提到 `stype/bib-style` 选 `〈其他样式〉` 且 `bib-backend = bibtex` 时要保证_相应的 .bst 文件_ 能被调用。所以正解是把参考文献的格式改为用 .bst 文件控制吗？如果这样的话要怎么操作呢？（比如 `stype/bib-style` 要设为什么）

修改 bst 文件是 bibtex 的方法，参考 `texdoc gbt7714`。

按以下步骤修改：
1. 命令行输入 `kpsewhich gbt7714-numerical.bst`
2. 将返回路径代表的文件复制到工作目录下
3. 用文本编辑器打开复制后的 `gbt7714-numerical.bst`
4. 找到 `uppercase.name` 项，将其 `#1` 改为 `#0`，删除辅助文件再重新编译。


关于参考文献的定制，我后面开一个专门讨论的地方吧，以便方法可以复用。

**Reply by iTomxy** [reply]
按第一种方法改用 biblatex 之后确实可以了，但连续引用的压缩效果没有了（`[1, 2, 3]` **不**会变成 `[1-3]`），无论 latexmk 还是 xelatex + biber 都是。按 [1] 说的在导言区那把 `citestyle = numeric-comp` 或 `style = numeric-comp` 也一齐传给 biblatex 宏包也**不**行，但是换回 bibtex 又可以压缩了，但人名样式没改。

现在是用第二种 .bst 的方法。除了你提到的步骤，根据 `texdoc gbt7714` 命令弹出来的那个 pdf 说明，还需要在导言区加：

```latex
\usepackage[sort&compress]{gbt7714}
\bibliographystyle{gbt7714-numerical}
%...
\begin{document}
%...
```

另外，在试的时候出现过引用文献爆行宽的情况，参照 [2]，加了：

```latex
\usepackage{blindtext}
%...
\begin{document}
\begin{sloppypar}
%...
\end{sloppypar}
\end{document}
```

---

1. [biblatex with numeric-comp doesn't compress](https://tex.stackexchange.com/questions/303986/biblatex-with-numeric-comp-doesnt-compress)
2. [texlive or latex 文字超出边界的处理方法 “Overfull \hbox (pt too wide)”](https://blog.csdn.net/u010801696/article/details/78868516)

**Reply by sikouhjw** [reply]
> 按第一种方法改用 biblatex 之后确实可以了，但连续引用的压缩效果没有了（`[1, 2, 3]` **不**会变成 `[1-3]`），无论 latexmk 还是 xelatex + biber 都是。按 [1] 说的在导言区那把 `citestyle = numeric-comp` 或 `style = numeric-comp` 也一齐传给 biblatex 宏包也**不**行，但是换回 bibtex 又可以压缩了，但人名样式没改。

我怀疑是你那边的问题，你之前的修改 cls 有弄回去吗？
```tex
\documentclass{gdutthesis}

\gdutsetup{
  style/bib-resource    = {gdutthesis-template.bib},
  style/bib-backend     = {biblatex},
}

\PassOptionsToPackage{gbnamefmt=lowercase}{biblatex}

\begin{document}
\cite{chendengyuan2000guoshi,woerdelun2012jingji}
\cite{chendengyuan2000guoshi,woerdelun2012jingji,beijing2012lidai}
\cite{chendengyuan2000guoshi,woerdelun2012jingji,beijing2012lidai,quanguo2010doubai}

\printbibliography
\end{document}
```
![图片](https://user-images.githubusercontent.com/49770925/162578796-f567b0a1-17a1-4019-8e43-575db5400798.png)


> 现在是用第二种 .bst 的方法。除了你提到的步骤，根据 `texdoc gbt7714` 命令弹出来的那个 pdf 说明，还需要在导言区加：
> 
> ```
> \usepackage[sort&compress]{gbt7714}
> \bibliographystyle{gbt7714-numerical}
> %...
> \begin{document}
> %...
> ```

不需要加这个，当使用 bibtex 的时候，调用的就是 `gbt7714`。

> 另外，在试的时候出现过引用文献爆行宽的情况，参照 [2]，加了：
> 
> ```
> \usepackage{blindtext}
> %...
> \begin{document}
> \begin{sloppypar}
> %...
> \end{sloppypar}
> \end{document}
> ```

这个是个问题，我没怎么管过……

**Reply by sikouhjw** [reply]
`\sloppy` 确实可以加到模板里。

**Reply by iTomxy** [reply]
哦确实，bibtex 不用加那个 gbt7714 包。

我用 git 恢复了 .cls 的内容，还是不行，不知道怎么回事…先用着 bibtex 的方案好了

**Reply by iTomxy** [reply]
我知道原因了，新 .cls 里面有 `\newtheorem{assumption}{假设}`，可能跟我之前自己打的冲突了，我把导言区自己打那句删了就可以适应更新过的 .cls 文件了

**Reply by sikouhjw** [reply]
debug 最好用命令行编译，这样报错明显点。
